File: SelfStudy/DoubleDispatch/doubleDispatch.swift
import Foundation

final class Processor {
  func process(_ obj: BaseClass) { 
    print("got a base object")
    obj.myFunction()
  }

  func process(_ obj: Derived1) {
    print("got a derived1 object")
    obj.myFunction()
  }

  func process(_ obj: Derived2) {
    print("got a derived2 object")
    obj.myFunction()
  }
}

class BaseClass {
  func myFunction() { print("BaseClass myFunction called") }
  func process(with proc: Processor){ proc.process(self) }
}

class Derived1: BaseClass {
  override func myFunction() { print("Derived1 myFunction called") }
  override func process(with proc: Processor){ proc.process(self) }
}

class Derived2: BaseClass {
  override func myFunction() { print("Derived2 myFunction called") }
  override func process(with proc: Processor){ proc.process(self) }
}

class Derived3: BaseClass {
  override func myFunction() { print("Derived3 myFunction called") }
  override func process(with proc: Processor){ proc.process(self) }
}

let p: Processor = Processor()

//first set results
let bc: BaseClass = BaseClass()
let dc1: Derived1 = Derived1()
let dc2: Derived2 = Derived2()
let dc3: Derived3 = Derived3()
p.process(bc)
p.process(dc1)
p.process(dc2)
p.process(dc3)

print("\n\n")

//second set results
let bc_1: BaseClass = bc
let dc1_1: BaseClass = dc1
let dc2_1: BaseClass = dc2
let dc3_1: BaseClass = dc3
p.process(bc_1)
p.process(dc1_1)
p.process(dc2_1)
p.process(dc3_1)

print("\n\n")

bc_1.process(with: p)
dc1_1.process(with: p)
dc2_1.process(with: p)
dc3_1.process(with: p)

File: SelfStudy/DoubleDispatch/doubleDispatch.kt
open class BaseClass {
    open fun myFunction() {
        println("BaseClass myFunction called")
    }

    open fun process(proc: Processor) {
        proc.process(this)
    }
}

class Derived1 : BaseClass() {
    override fun myFunction() {
        println("Derived1 myFunction called")
    }

    override fun process(proc: Processor) {
        proc.process(this)
    }
}

class Derived2 : BaseClass() {
    override fun myFunction() {
        println("Derived2 myFunction called")
    }

    override fun process(proc: Processor) {
        proc.process(this)
    }
}

class Derived3 : BaseClass() {
    override fun myFunction() {
        println("Derived3 myFunction called")
    }

    override fun process(proc: Processor) {
        proc.process(this)
    }
}

class Processor {
    fun process(obj: BaseClass) {
        println("got a base object")
        obj.myFunction()
    }

    fun process(obj: Derived1) {
        println("got a derived1 object")
        obj.myFunction()
    }

    fun process(obj: Derived2) {
        println("got a derived2 object")
        obj.myFunction()
    }
}

fun main() {
    val p = Processor()

    val bc: BaseClass = BaseClass()
    val dc1: Derived1 = Derived1()
    val dc2: Derived2 = Derived2()
    val dc3: Derived3 = Derived3()

    p.process(bc)
    p.process(dc1)
    p.process(dc2)
    p.process(dc3)
    
    println("\n\n")
    
    val bc_1: BaseClass = bc
    val dc1_1: BaseClass = dc1
    val dc2_1: BaseClass = dc2
    val dc3_1: BaseClass = dc3
    
    p.process(bc_1)
    p.process(dc1_1)
    p.process(dc2_1)
    p.process(dc3_1)
    
    println("\n\n")
    
    bc_1.process(p)
    dc1_1.process(p)
    dc2_1.process(p)
    dc3_1.process(p)
}